trie-db/src/triedbmut.rs

```diff
@@ -634,65 +652,78 @@ impl<'a, L: TrieLayout> Index<&'a StorageHandle> for NodeStorage<L> {
 	fn index(&self, handle: &'a StorageHandle) -> &Node<L> {
 		match self.nodes[handle.0] {
 			Stored::New(ref node) => node,
-			Stored::Cached(ref node, _) => node,
+			Stored::Cached(ref node, _, _) => node,
 		}
 	}
 }
 
 /// A builder for creating a [`TrieDBMut`].
 pub struct TrieDBMutBuilder<'db, L: TrieLayout> {
-	db: &'db mut dyn HashDB<L::Hash, DBValue>,
-	root: &'db mut TrieHash<L>,
-	cache: Option<&'db mut dyn TrieCache<L::Codec>>,
-	recorder: Option<&'db mut dyn TrieRecorder<TrieHash<L>>>,
+	db: &'db dyn HashDB<L::Hash, DBValue, L::Location>,
+	root: TrieHash<L>,
+	root_location: L::Location,
+	cache: Option<&'db mut dyn TrieCache<L::Codec, L::Location>>,
+	recorder: Option<&'db mut dyn TrieRecorder<TrieHash<L>, L::Location>>,
 }
 
 impl<'db, L: TrieLayout> TrieDBMutBuilder<'db, L> {
 	/// Create a builder for constructing a new trie with the backing database `db` and empty
 	/// `root`.
-	pub fn new(db: &'db mut dyn HashDB<L::Hash, DBValue>, root: &'db mut TrieHash<L>) -> Self {
-		*root = L::Codec::hashed_null_node();
-
-		Self { root, db, cache: None, recorder: None }
+	pub fn new(db: &'db dyn HashDB<L::Hash, DBValue, L::Location>) -> Self {
+		let root = L::Codec::hashed_null_node();
+		Self { root, db, cache: None, recorder: None, root_location: Default::default() }
 	}
 
 	/// Create a builder for constructing a new trie with the backing database `db` and `root`.
 	///
 	/// This doesn't check if `root` exists in the given `db`. If `root` doesn't exist it will fail
 	/// when trying to lookup any key.
 	pub fn from_existing(
-		db: &'db mut dyn HashDB<L::Hash, DBValue>,
-		root: &'db mut TrieHash<L>,
+		db: &'db dyn HashDB<L::Hash, DBValue, L::Location>,
+		root: TrieHash<L>,
 	) -> Self {
-		Self { db, root, cache: None, recorder: None }
+		Self { db, root, cache: None, recorder: None, root_location: Default::default() }
+	}
+
+	/// Same as `from_existing` but force a db location to access root.
+	/// Note root in parameter is not checked.
+	pub fn from_existing_with_db_location(
+		db: &'db dyn HashDB<L::Hash, DBValue, L::Location>,
+		root: TrieHash<L>,
```

Comment:
(change from arkpar/new-trie)
Here, root is not really usefull, since we use location instead. Maybe something to do, either force a check or removing this parameter. (also true for triedbmut)